gobblin-core/src/main/java/gobblin/publisher/HiveRegistrationPublisher.java

```diff
@@ -92,23 +92,29 @@ public void publishData(Collection<? extends WorkUnitState> states) throws IOExc
     CompletionService<Collection<HiveSpec>> completionService =
         new ExecutorCompletionService<>(this.hivePolicyExecutor);
 
-    //addRuntimeHiveRegistrationProperties(super.state);
-    final HiveRegistrationPolicy policy = HiveRegistrationPolicyBase.getPolicy(super.state);
+    final State superState = super.state;
+    int numberOfPathsToRegister = getNumberOfPathsToRegister(states);
 
-    Set<String> pathsToRegister = getUniquePathsToRegister(states);
-    log.info("Number of paths to be registered in Hive: " + pathsToRegister.size());
-    for (final String path : pathsToRegister) {
-      completionService.submit(new Callable<Collection<HiveSpec>>() {
-
-        @Override
-        public Collection<HiveSpec> call() throws Exception {
-          return policy.getHiveSpecs(new Path(path));
+    // Each state in states is task-level State, while superState is the Job-level State.
+    // Using both State objects to distinguish each HiveRegistrationPolicy so that
+    // they can carry task-level information to pass into Hive Partition and its corresponding Hive Table.
+    for (State state:states) {
```

Comment:
Previously it generates all Paths to register through one function call and all Paths are registered based on a single policy. 
The code change here is trying to iterate through all states and create different policy objects for state, so that different policy object can carry the topic information for each path to register.